### PR TITLE
[MNT-24528] Add aspects without any properties with API using multipart/form-data

### PR DESCRIPTION
**Pull request by jottley**
An enhancement request to allow customers to create nodes with both content AND aspectNames in a single REST API call. Currently, the Alfresco REST API has two separate paths:

1. **JSON body (createNode)**: Can set aspectNames but CANNOT upload binary content
2. **Multipart/form-data (upload)**: Can upload binary content but does NOT support aspectNames

This forces customers to make two API calls: one to upload content, another to add aspects. Customers want to enable aspectNames in the multipart upload endpoint. This should be done without breaking existing API behavior.

### Design Decisions

1. **Field Name**: Use `"aspectnames"` (lowercase) to match JSON API naming pattern (`aspectNames`) and existing multipart conventions (all lowercase field names)
2. **Helper Method**: Create new `parseCommaSeparatedList()` instead of renaming existing `getRequestedRenditions()` to avoid breaking changes
3. **Return Type**: Use `List<String>` to match existing method signatures
4. **Validation Order**: Validate aspects BEFORE path creation, following the properties validation pattern
5. **Aspect Addition**: Add aspects AFTER node creation, following the createNode() pattern

#### 1. Add a Helper Method (NodesImpl.java ~line 3632)

Add after the existing `getRequestedRenditions()` method:

```java
/**
 * Parse a comma-separated list of names into a List.
 * 
 * @param namesParam comma-separated string (e.g., "cm:titled,cm:author")
 * @return List of trimmed, non-empty names, or an empty list if input is null
 */
static List<String> parseCommaSeparatedList(String namesParam)
{
    if (namesParam == null)
    {
        return Collections.emptyList();;
    }

    String[] nameArray = namesParam.split(",");
    List<String> names = new ArrayList<>(nameArray.length);
    
    for (String name : nameArray)
    {
        name = name.trim();
        if (!name.isEmpty())
        {
            names.add(name);
        }
    }
    
    return names.isEmpty() ? Collections.emptyList(); : names;
}
```

#### 2. Add Variable Declaration (NodesImpl.java line 3378)

After `String renditionNames = null;`:

```java
String aspectNames = null;
```

#### 3. Add Switch Case (NodesImpl.java ~line 3435)

After the "renditions" case, before "versioningenabled" case:

```java
case "aspectnames":
    aspectNames = getStringOrNull(field.getValue());
    break;
```

#### 4. Parse and Validate Aspects (NodesImpl.java line 3486)

BEFORE the existing `validateProperties()` line, add:

```java
final List<String> aspects = parseCommaSeparatedList(aspectNames);
validateAspects(aspects, EXCLUDED_NS, EXCLUDED_ASPECTS);
```

#### 5. Add Aspects to Node (NodesImpl.java ~line 3530)

RIGHT AFTER `NodeRef nodeRef = createNewFile(...);` and BEFORE `final Node fileNode = getFolderOrDocumentFullInfo(...)`:

```java
addCustomAspects(nodeRef, aspects, EXCLUDED_ASPECTS);
```

#### 6. Update Test Infrastructure (MultiPartBuilder.java line 286)

Change from:
```java
addPartIfNotNull(parts, "aspects", getCommaSeparated(aspects));
```

To:
```java
addPartIfNotNull(parts, "aspectnames", getCommaSeparated(aspects));
```

## Edge Cases Handled

- **Null/empty aspectNames**: Returns an empty list from parseCommaSeparatedList(), handled gracefully by addCustomAspects()
- **Invalid aspect names**: `validateAspects()` → `mapToNodeAspects()` throws InvalidArgumentException
- **Excluded aspects**: validateAspects() checks EXCLUDED_ASPECTS list
- **Excluded namespaces** (sys:*): validateAspects() checks EXCLUDED_NS list
- **Whitespace in names**: Trimmed by parseCommaSeparatedList()
- **Duplicate aspects**: Safely handled by nodeService.addAspect()
- **Overwrite scenario**: When overwrite=true, code uses updateExistingFile() which doesn't call createNewFile(), so aspects won't be added (consistent with current behavior - aspects would need separate PUT call)

### Backward Compatibility: ✅ NO BREAKING CHANGES

- Adding new optional parameter `aspectnames` to multipart upload
- Existing calls without aspectnames work exactly as before
- No changes to response format (aspectNames already included in response)
- No changes to existing method signatures visible to other classes

### Forward Compatibility: ✅ FUTURE-PROOF

- Follows established patterns from createNode() method
- Uses existing validation and aspect addition methods
- Consistent with JSON API field naming conventions
- Clean separation allows future enhancements

### Test Coverage Required

1. **Basic Upload with Single Aspect**
   ```java
   MultiPartBuilder.create()
       .setFileData(new FileData("test.txt", file))
       .setAspects(Arrays.asList("cm:titled"))
       .build()
   ```
   - Assert: Response includes "cm:titled" in aspectNames
   - Assert: GET node confirms aspect is present

2. **Upload with Multiple Aspects**
   - Test: Arrays.asList("cm:titled", "cm:author")
   - Verify: Both aspects present on node

3. **Null/Empty Aspects** (should succeed)
   - Test: .setAspects(null)
   - Test: .setAspects(Collections.emptyList())

4. **Invalid Aspect Name** (should return 400)
   - Test: .setAspects(Arrays.asList("invalid:aspect"))
   - Expect: InvalidArgumentException

5. **Excluded Namespace** (should return 400)
   - Test: .setAspects(Arrays.asList("sys:referenceable"))
   - Expect: IllegalArgumentException

6. **Whitespace Handling**
   - Test: Manual form-data with "cm:titled, cm:author  "
   - Verify: Properly trimmed

7. **Combined Parameters Test**
   - Test: aspectNames + relativePath + autoRename + renditions
   - Verify: All parameters work together

8. **Overwrite Scenario**
   - Test: Upload with aspectNames, then overwrite=true
   - Verify: Overwrite uses updateExistingFile() path (no aspect addition)

### Manual Testing

```bash
# Test multipart upload with aspectNames
curl -X POST \
  'http://localhost:8080/alfresco/api/-default-/public/alfresco/versions/1/nodes/{parentId}/children' \
  -H 'Authorization: Basic ...' \
  -F 'filedata=@test.pdf' \
  -F 'name=test.pdf' \
  -F 'aspectnames=cm:titled,cm:author' \
  -F 'cm:title=My Document Title'

# Verify aspectNames in response
{
  "id": "...",
  "name": "test.pdf",
  "aspectNames": ["cm:auditable", "cm:titled", "cm:author", ...],
  "properties": {
    "cm:title": "My Document Title"
  }
}
```

### Low Risk Changes
- ✅ Purely additive feature (no existing behavior modified)
- ✅ Uses proven validation methods (validateAspects, addCustomAspects)
- ✅ Follows existing patterns from createNode() implementation
- ✅ Existing security checks (EXCLUDED_NS, EXCLUDED_ASPECTS) enforced

### Minimal Performance Impact
- Adding aspects is an existing operation, just exposed via new parameter
- No additional database queries beyond what createNode() already does
- Validation happens before node creation (fail fast)

### Zero Security Concerns
- Same validation as JSON API path
- EXCLUDED_NS prevents system namespace manipulation
- EXCLUDED_ASPECTS list enforced
- cm:auditable automatically filtered out

## Summary

This change:

1. **Fixes all issues** in the proposed change
2. **Avoids breaking changes** by creating new helper method
3. **Maintains consistency** with JSON API and existing patterns
4. **Includes comprehensive tests** for all edge cases
5. **Preserves backward compatibility** (purely additive)

### CODE REVIEW

File: remote-api/src/test/java/org/alfresco/rest/api/tests/NodeApiTest.java [hunk 1]
```diff
@@ -6635,4 +6635,247 @@ private void setSiteNodeInheritance(String siteId, String nodeId, boolean inheri
...
+                .setAspects(Arrays.asList("cm:titled", "cm:author"));
+        MultiPartRequest reqBody = multiPartBuilder.build();
+
+        HttpResponse response = post(getNodeChildrenUrl(folderId), reqBody.getBody(), null, reqBody.getContentType(), 201);
+        Document document = RestApiUtil.parseRestApiEntry(response.getJsonResponse(), Document.class);
+
+        assertEquals(fileName, document.getName());
+        assertTrue("Document should have cm:titled aspect", document.getAspectNames().contains("cm:titled"));
+        assertTrue("Document should have cm:author aspect", document.getAspectNames().contains("cm:author"));
+    }
```
[the remark this answers is not in the document]
Comment (jottley) [reply remote-api/src/test/java/org/alfresco/rest/api/tests/NodeApiTest.java:6843]: @copilot apply changes based on [this feedback](https://github.com/Alfresco/alfresco-community-repo/pull/4042#discussion_r3120072957)

File: remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java [hunk 2]
```diff
@@ -3483,6 +3489,8 @@ public Node upload(String parentFolderNodeId, FormData formData, Parameters para
         final QName assocTypeQName = ContentModel.ASSOC_CONTAINS;
         final Set<String> renditions = getRequestedRenditions(renditionNames);
 
+        final List<String> aspects = parseCommaSeparatedList(aspectNames);
+        validateAspects(aspects, EXCLUDED_NS, EXCLUDED_ASPECTS);
         validateProperties(qnameStrProps, EXCLUDED_NS, Arrays.asList());
```
[the remark this answers is not in the document]
Comment (jottley) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java:3494]: @copilot apply changes based on [this feedback](https://github.com/Alfresco/alfresco-community-repo/pull/4042#discussion_r3120072986)

**Review by evasques — APPROVED** [review]

**Review by pzhyland — APPROVED** [review]
Great! Please remember about the https://github.com/Alfresco/rest-api-explorer

**Review by tiagosalvado10 — APPROVED** [review]
